README.md
![example workflow name](https://github.com/haikyuu/inertia-node-adapter/workflows/test-and-build/badge.svg)

Note that while this is working software. It's still in very early experimentation stage.

# inertia-node-adapter

Inertia adapter for nodejs - supports Express and Koa.

This is a fork of https://github.com/jordankaerim/inertia-node with added support for koa, flash messages and integration tests that adhere to Inertia protocol.

This also includes a flash middleware for koa and express as a bonus. You can use it or use your own.

## Usage

For working examples, see the `__tests/testServers` folders. It contains two servers in koa and express. Note The code is a bit tangled because it was written for integration tests in mind.

### Koa

```typescript
import { inertiaKoaAdapter as inertia, koaFlash } from 'inertia-node-adapter';
import session from 'koa-session';
import Router from '@koa/router';

const version = '1';

app.use(session({ maxAge: 86400000 }, app));
app.use(koaFlash()); // optional, but requires using middleware session beforehand
app.use(
  inertia({
    version,
    html: getHtml,
    flashMessages: (ctx) => ctx.flash.flashAll(), // optional, depends on the flash middleware used
  })
);
const router = new Router();
// in your router handlers
router.get('/home', async (ctx, next) => {
  // if flash middleware was added.
  ctx.flash.setFlashMessage('success', 'User created successfully');

  ctx.Inertia.render({
    component: 'home',
    props: { name: 1 },
    url: '/home',
    version,
  });
  next();
})
  app.use(router.routes());
  return next();
});
```

### Express usage

```typescript
import express from 'express';
import session from 'express-session';

const app = express();

app.use(session({ secret: 'secret' }));
const router = express.Router();

router.get('/home', async (req, res, next) => {
  // if flash middleware was added.
  req.flash.setFlashMessage('success', 'User created successfully');

  req.Inertia.render({
    component: 'home',
    props: { name: 1 },
    url: '/home',
    version,
  });
});
app.use(flash());
app.use(
  inertia({
    version,
    html: getHtml,
    flashMessages: (req) => req.flash.flashAll() },
  })
);
app.use(router);
```

## Usage details

`inertia-node` expects two arguments:

1. `html`: a function that receives the Inertia `page` object and an optional `viewData` object that you can populate with additional data. The function should return an HTML string.
2. `version` (optional): your current asset version

It will return a standard Node.js middleware that you can use with Express.js, Polka etc (or koa if you're using koa adapter). Functions can be accessed from the `Inertia` object that will be added to the `request` / `context`. You can chain functions together but you can't call another function after calling [`render`](#renderPage) or [`redirect`](#redirectUrl).

**Note:** In your HTML view function make sure to always stringify the `page` object and include it in the `data-page` attribute of the HTML node you want to render your JavaScript app in. For more infos on how Inertia works read [the protocol](https://inertiajs.com/the-protocol) on the Inertia website.

## API

### setViewData(data)

1. `data`, _Object_ - An Object of additional data you want to pass to your HTML view function

`setViewData` can be used to pass additional data to your [HTML view function](#usage) such as the page's title or other meta data.

```javascript
.use(({ Inertia }, _, next) => {
  Inertia.setViewData({
    title: "Todo App",
    description:"A Web App to Create and Manage To-dos"
  })

  next();
})

// ...

const html = (page, viewData) => `
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width,initial-scale=1" />

    <meta name="description" content="${viewData.description}">
    <title>${viewData.title}</title>

    <link rel="stylesheet" href="/build/bundle.css" />
    <script defer type="module" src="/build/main.js"></script>
  </head>

  <body id="app" data-page='${JSON.stringify(page)}'></body>
</html>
`;
```

### shareProps(props)

1. `props`, _Object_ - An object of props you want to include with every `render` call

Shared props are props that will be combined with the props you pass to the [`render`](#renderPage) function. When you call `shareProps` more than once, props shared from previous calls will be merged with props from any subsequent calls.

```javascript
.use(({ Inertia }, _, next) => {
  Inertia.shareProps({ username: "ironMan" })
  next();
})
.get("/todo", ({ Inertia }) => {
  Inertia.render({
    component: "Todo",
    props: {
      todo: [
        { text: "Study Korean", done: false },
        { text: "Cook Arabic food", done: false },
      ]
    }
    // JavaScript component on the client will receive
    // {
    //   username: "ironMan",
    //   todo: [
    //     { "Study Korean", done: false },
    //     { "Cook Arabic food", done: false },
    //   ]
    // }
  });
})
```

### setHeaders(headers)

1. `headers`, _Object_ - An Object of custom headers you want to include in your response

Add custom headers to your response.

**Note:** Headers that are required by Inertia take precedence and cannot be overwritten.

```javascript
.get("/", ({ Inertia }) => {
  Inertia
    .setHeaders({
      token: "7pTgHCv0JgeAyyBRDpUi"
    })
    .render({
      component: "Index",
      props: { username: "ironMan" }
    });
})
```

### setStatusCode(statusCode)

1. `statusCode`, _number_ - The response's status code

Change the status code when sending a response. Useful for e.g. when you want to render an error. Headers set from previous calls will be merged with headers set from any subsequent calls.

```javascript
.get("/", ({ Inertia }) => {
  Inertia.render({
    component: "Index",
    props: { username: "ironMan" }
  });
})
.use(({ Inertia }) => {
  Inertia
    .setStatusCode(404)
    .render({
      component: "Error",
      props: { message: "Page not found" },
    });
})
```

### render(page)

1. `page`, _Object_ - the Inertia page object

This function will send your response as either a HTML or JSON string to the client depending on wether the client is requesting your page for the first time or is making a subsequent Inertia request.

The Inertia page object consists of the following properties:

1. `component`, _string_ - The name of the JavaScript component to render on the client
2. `props`, _Object_ - The page props (data)
3. `url`, _string_ - The URL of the route. This will be automatically added by the middleware.
4. `version`, _string_ - The asset version. This will be automatically added by the middleware.

```javascript
.get("/todo", ({ Inertia }) => {
  Inertia.render({
    component: "Todo",
    props: {
      todo: [
        { text: "Study Korean", done: false },
        { text: "Cook Arabic food", done: false },
      ]
    },
  });
})
```

On Partial Reloads only props requested by the client will be sent. To improve performance on the server you can wrap each prop inside a function so that they will only be evaluated when necessary.

```javascript
.get("/todo", ({ Inertia }) => {
  const todoArray = async () => await database.getTodo();
  const bookmarks = async () => await database.getBookmarks();

  Inertia.render({
    component: "Todo",
    props: {
      todoArray,
      bookmarks
    },
  });
})
```

Now when the client only requests `todoArray` the middleware will hit the database only once and not call `bookmarks`.

### redirect(url)

1. `url`, _string_ - The URL to redirect to

Redirect the client to a different URL.

```javascript
.post("/todo", (req, res) => {
  database.createTodo(req.body);
  req.Inertia.redirect("/todo");
})
```

**Note:** Inertia requires you to use a `303` status code when redirecting upon a `PUT`, `PATCH` or `DELETE` request and a `302` status code otherwise. The `redirect` function will automatically take care of this for you. If you handle redirects yourself make sure to select the correct status code.

**Note:** Calling [`setStatusCode`](#renderPage) before [`redirect`](#redirectUrl) has no effect.

## Credits

- [Jordan Kaerim](https://github.com/jordankaerim).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.0.4](https://github.com/haikyuu/inertia-node-adapter/compare/v1.0.3...v1.0.4) (2021-01-18)

### [1.0.3](https://github.com/haikyuu/inertia-node-adapter/compare/v1.0.2...v1.0.3) (2021-01-11)


### Bug Fixes

* **index:** fix default exports ([56f5e0b](https://github.com/haikyuu/inertia-node-adapter/commit/56f5e0bea04a79e2bb1a5269af288c32488ba3db))

### 1.0.2 (2021-01-11)

### 1.0.1 (2021-01-11)


src/__tests__/koa.folio.ts
// koa.folio.ts
import { Server } from 'http';

import { folio as base } from 'folio';
import { DefaultState } from 'koa';
import sinon, { SinonSandbox, SinonSpy } from 'sinon';

import { Page } from '../lib/koaAdapter';

import getServer from './testServers';

// Declare worker fixtures.
type koaWorkerFixtures = {
  port: number;
  koa: {
    server: Server;
    metadata: {
      fakes: { [key: string]: SinonSpy };
      pages: { [key: string]: Page };
    };
    sandbox: SinonSandbox;
  };
};
export type serverType = 'koa' | 'express';
const fixtures = base.extend<
  DefaultState,
  koaWorkerFixtures,
  { type: serverType }
>();

// |port| fixture has a unique value value of the worker process index.
fixtures.port.init(
  async ({ testWorkerIndex }, run) => {
    await run(3005 + testWorkerIndex);
  },
  { scope: 'worker' }
);

fixtures.type.initParameter('API version', 'koa');

// |koa| fixture starts automatically for every worker.
fixtures.koa.init(
  async ({ port, type }, run) => {
    const newSandbox = sinon.createSandbox();
    const { app, metadata, sandbox } = getServer(type, { sandbox: newSandbox });
    let server: ReturnType<typeof app.listen> = new Server(); // hack to fix typescript complaining about use before assigned
    console.log('Starting server...');
    await new Promise((f: (value: unknown) => void) => {
      server = app.listen(port, () => f(1));
    });
    console.log('Server ready');

    await run({ server, metadata, sandbox });
    console.log('Stopping server...');
    await new Promise((f) => server.close(f));
    console.log('Server stopped');
  },
  { scope: 'worker', auto: true }
);

const { it, describe, expect, beforeEach, afterEach } = fixtures.build();

export { it, describe, expect, beforeEach, afterEach };


src/__tests__/testServers/koa.ts
import path from 'path';

import Router from '@koa/router';
import stringify from 'json-stable-stringify';
import Koa, { DefaultContext } from 'koa';
import session from 'koa-session';
import serve from 'koa-static';

import inertia, { Inertia, Page } from '../../lib/koaAdapter';
import flash, { Flash } from '../../lib/koaFlash';

import { ServerOptions } from '.';

export function getHtml(page: Page) {
  return `<!DOCTYPE html><html dir="ltr" lang="en">
<head><meta charset="utf-8"><title>Inertia adapter</title></head>
<body><div data-page='${stringify(page)}'></div>
</body></html>`;
}
const version = '1';

export function koaServer({ sandbox }: ServerOptions) {
  const app = new Koa();

  app.keys = ['some secret'];
  app.use(
    session(
      {
        maxAge: 86400000,
      },
      app
    )
  );
  const fakes = {
    numbers: sandbox.fake.resolves([1, 2, 3]),
    bigNumbers: sandbox.fake.resolves([44, 33, 1123]),
  };
  const pages = {
    index: { component: 'index', props: { name: 'SERVER' }, url: '/', version },
    home: { component: 'app-home', props: { name: 1 }, url: '/home', version },
    partial: {
      component: 'partial-component',
      version,
      url: '/partial-optimized',
      props: {
        numbers: fakes.numbers,
        bigNumbers: fakes.bigNumbers,
      },
    },
  };
  const metadata = { fakes, pages };

  type FlashMessages = 'info' | 'error' | 'success';
  const router = new Router<
    DefaultContext,
    { Inertia: Inertia; flash: Flash<FlashMessages> }
  >();

  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore this fixes a bug with session trying to access app.keys using this.keys
  router.keys = app.keys;
  router.get('/flash', async (ctx, next) => {
    // set session data
    ctx.flash.setFlashMessage('success', 'User created successfully');
    ctx.Inertia.render(pages.home);
    return next();
  });
  router.get('/home', async (ctx, next) => {
    ctx.Inertia.render(pages.home);
    return next();
  });

  router.post('/home', async (ctx, next) => {
    ctx.Inertia.render(pages.home);
    return next();
  });

  router.get('/partial-optimized', async (ctx, next) => {
    if (ctx.query?.redirect) {
      await ctx.Inertia.render(pages.home);
    } else {
      await ctx.Inertia.render(pages.partial);
    }
    return next();
  });
  app.use(flash<FlashMessages>());
  app.use(
    serve(path.join(__dirname, '..', '..', 'www'), { brotli: true, gzip: true })
  );
  app.use(
    inertia({
      version,
      html: getHtml,
      flashMessages: (ctx) => ctx.flash.flashAll(),
    })
  );

  app.use(router.routes());
  app.use(router.allowedMethods());
  return { app, metadata, sandbox };
}


src/__tests__/inertia-spec.test.ts
import { TestModifier } from 'folio/out/testModifier';
import fetch from 'node-fetch';

import { headers } from '../lib/koaAdapter';

import { afterEach, describe, expect, it } from './koa.folio';

function skip(skip = true) {
  return (test: TestModifier) => test.skip(skip, 'Skip');
}

const cookies = {
  koa: 'koa.sess',
  express: 'connect.sid',
};
describe('Inertia adapter', () => {
  afterEach(async ({ koa }) => {
    koa.sandbox.reset();
  });

  it('responds with the html page provided in initialization if no inertia header is there', async ({
    port,
  }) => {
    const result = await fetch(`http://localhost:${port}/home`, {
      headers: {},
    });
    const html = await result.text();
    expect(html).toMatchSnapshot();
  });

  it('responds with a json response with encoded page object when inertia header is present and version is the same', async ({
    port,
    koa,
  }) => {
    const result = await fetch(`http://localhost:${port}/home`, {
      headers: { 'X-Inertia': 'true', 'X-Inertia-Version': '1' },
    });
    expect(await result.json()).toMatchObject(
      koa.metadata.pages.home as Record<string, unknown>
    );
  });

  it('includes partial data headers on partial request', async ({
    port,
    koa,
  }) => {
    const result = await fetch(`http://localhost:${port}/partial-optimized`, {
      headers: {
        'X-Inertia': 'true',
        'X-Inertia-Version': '1',
        'X-Inertia-Partial-Data': 'numbers',
        'X-Inertia-Partial-Component': 'partial-component',
      },
    });
    expect(result.status).toBe(200);
    expect(await result.json()).toMatchObject({
      ...koa.metadata.pages.partial,
      props: { numbers: [1, 2, 3] },
    });
    // we are not resetting fake function calls. Might fail
    expect(koa.metadata.fakes.bigNumbers.callCount).toBe(0);
    expect(koa.metadata.fakes.numbers.callCount).toBe(1);
  });
  it('partial requests without the name of the component return a full request', async ({
    port,
    koa,
  }) => {
    const result = await fetch(`http://localhost:${port}/partial-optimized`, {
      headers: {
        'X-Inertia': 'true',
        'X-Inertia-Version': '1',
        'X-Inertia-Partial-Data': 'numbers',
      },
    });
    expect(result.status).toBe(200);
    expect(await result.json()).toMatchObject({
      ...koa.metadata.pages.partial,
      props: { numbers: [1, 2, 3] },
    });
    expect(koa.metadata.fakes.bigNumbers.callCount).toBe(1);
    expect(koa.metadata.fakes.numbers.callCount).toBe(1);
  });
  it(
    'includes all data headers on none partial request',
    skip(false),
    async ({ port, koa }) => {
      const result = await fetch(`http://localhost:${port}/partial-optimized`, {
        headers: {
          'X-Inertia': 'true',
          'X-Inertia-Version': '1',
        },
      });
      expect(result.status).toBe(200);
      expect(await result.json()).toMatchObject({
        ...koa.metadata.pages.partial,
        props: { numbers: [1, 2, 3], bigNumbers: [44, 33, 1123] },
      });
      expect(koa.metadata.fakes.bigNumbers.callCount).toBe(1);
      expect(koa.metadata.fakes.numbers.callCount).toBe(1);
    }
  );

  it('flash messages are shared as props', async ({ port, koa, type }) => {
    const result = await fetch(`http://localhost:${port}/flash`, {
      headers: { 'X-Inertia': 'true', 'X-Inertia-Version': '1' },
    });
    // flash messages are stored in a cookie
    expect(result.headers.get('set-cookie')?.startsWith(cookies[type])).toBe(
      true
    );
    const jsonResult = await result.json();
    expect(jsonResult).toMatchObject({
      ...koa.metadata.pages.home,
      url: '/flash',
      props: {
        ...koa.metadata.pages.home.props,
        success: ['User created successfully'],
      },
    });
  });
  it('re-flashes data when flash session data exists when a 409 Conflict', async ({
    port,
  }) => {
    const result = await fetch(`http://localhost:${port}/flash`, {
      headers: { 'X-Inertia': 'true', 'X-Inertia-Version': '2' },
    });
    expect(result.status).toBe(409);
    expect(result.statusText).toBe('Conflict');
    expect(result.headers.get(headers.xInertiaLocation)).toBe(`/flash`);
    expect(result.headers.get('set-cookie')).toBe(null);
    // no flash messages
  });

  it(
    'responds with 409 conflict when inertia header is present and version is different',
    skip(false),
    async ({ port }) => {
      const result = await fetch(`http://localhost:${port}/home`, {
        headers: { 'X-Inertia': 'true', 'X-Inertia-Version': '2' },
      });
      expect(result.status).toBe(409);
      expect(result.statusText).toBe('Conflict');
      expect(result.headers.get(headers.xInertiaLocation)).toBe(`/home`);
    }
  );
  it('Note, `409 Conflict` responses are only sent for `GET` requests, and not for POST/PUT/PATCH/DELETE requests', async ({
    port,
  }) => {
    const result = await fetch(`http://localhost:${port}/home`, {
      body: '12',
      method: 'POST',
      headers: { 'X-Inertia': 'true', 'X-Inertia-Version': '2' },
    });
    expect(result.status).toBe(200);
  });
  it('Partial Reload If the final destination is different for whatever reason, no partial reload occurs', async ({
    koa,
    port,
  }) => {
    const result = await fetch(
      `http://localhost:${port}/partial-optimized?redirect=home`,
      {
        headers: {
          'X-Inertia': 'true',
          'X-Inertia-Version': '1',
          'X-Inertia-Partial-Data': 'numbers',
        },
      }
    );
    expect(result.status).toBe(200);
    expect(await result.json()).toMatchObject({
      ...koa.metadata.pages.home,
      url: '/partial-optimized?redirect=home',
    });
    expect(koa.metadata.fakes.bigNumbers.callCount).toBe(0);
    expect(koa.metadata.fakes.numbers.callCount).toBe(0);
  });
});


src/lib/koaFlash.ts
/* eslint-disable @typescript-eslint/no-non-null-assertion */
import { DefaultContext, Middleware, Next, ParameterizedContext } from 'koa';

export interface Flash<FlashMessageType extends string = string> {
  setFlashMessage: (type: FlashMessageType, message: string) => number;
  setFlashMessages: (type: FlashMessageType, message: string[]) => number;
  flash: (type: FlashMessageType) => string[];
  flashAll: () => Record<FlashMessageType, string[]>;
}

export type FlashMiddleware = <FlashMessageType extends string>() => Middleware<
  ParameterizedContext<{ flash: Flash<FlashMessageType> }, DefaultContext>
>;

/**
 * Registers a middleware responsible for managing flash messages.
 * You can add a generic to the call to define the types of flash messages in your app
 * @example app.use(flash<"error" | "info" | "success">())
 *
 * @note It needs a session middleware in order to work. Has been tested with koa-session
 */
const flash: FlashMiddleware = <FlashMessageType extends string>() => {
  return async (ctx, next: Next) => {
    if (!ctx.session) {
      throw new Error(
        'Koa flash requires a session middleware to be added before'
      );
    }
    type MessagesRecord = Record<FlashMessageType, string[]>;
    class flash {
      /**
       * Appends the message given to the flash messages in the session of that type.
       * Creates a new array if the type isn't defined in the session flash messages.
       * @example
       * ctx.flash.setFlashMessage("info", "message sent")
       * ctx.session.flashMessages.info // now contains ["old message", "message sent"]
       * @param type the type of the flash message. picked from the generic if defined. Falls back to string
       * @param message the flash message
       */
      setFlashMessage(type: FlashMessageType, message: string) {
        const messagesRecord: MessagesRecord = (ctx.session!.flashMessages =
          ctx.session!.flashMessages || {});

        if (messagesRecord[type]) {
          messagesRecord[type].push(message);
        } else {
          messagesRecord[type] = [message];
        }
        return messagesRecord[type].length;
      }
      /**
       * Appends multiple messages to the flash messages in the session of that type.
       * Creates a new array if the type isn't defined in the session flash messages.
       * @example
       * ctx.flash.setFlashMessages("info", ["message sent", "user created"])
       * ctx.session.flashMessages.info // now contains ["old message", "message sent", "user created"]
       * @param type the type of the flash message. picked from the generic if defined. Falls back to string
       * @param messages  the flash messages
       */
      setFlashMessages(type: FlashMessageType, messages: string[]) {
        const messagesRecord: MessagesRecord = (ctx.session!.flashMessages =
          ctx.session!.flashMessages || {});
        if (!messagesRecord[type]) {
          messagesRecord[type] = messages;
        } else {
          messages.forEach((singleMessage) => {
            messagesRecord[type].push(singleMessage);
          });
        }
        return messagesRecord[type].length;
      }
      /**
       * Purges the given type messages from the session and returns the messages purged in an array
       * @example
       * ctx.body = { errors: ctx.flash.flash("error") } // {errors: ["upload failed", "server down"]}
       * @param type the type of the flash message. picked from the generic if defined. Falls back to string
       */
      flash(type: FlashMessageType) {
        const messagesRecord: MessagesRecord = (ctx.session!.flashMessages =
          ctx.session!.flashMessages || {});
        const msgs = messagesRecord[type];
        delete messagesRecord[type];
        return msgs;
      }

      /**
       * Purges the all messages from the session and returns the messages purged in a record
       * @example
       * ctx.body = ctx.flash.flash() // { errors: ["upload failed", "server down"], info: ["internet is fast"] }
       * @param type the type of the flash message. picked from the generic if defined. Falls back to string
       */
      flashAll() {
        const messagesRecord: MessagesRecord = ctx.session!.flashMessages || {};

        ctx.session!.flashMessages = {};
        return messagesRecord;
      }
    }
    ctx.flash = new flash();
    return next();
  };
};
export default flash;


src/lib/koaAdapter.ts
import debug from 'debug';
import type {
  Context,
  DefaultContext,
  DefaultState,
  Middleware,
  Next,
  ParameterizedContext,
} from 'koa';

const log = debug('inertia-node-adapter:koa');

type props = Record<string | number | symbol, unknown>;
export type Options = {
  readonly version: string;
  readonly html: (page: Page, viewData: props) => string;
  readonly flashMessages?: (
    context: DefaultContext & Record<string, unknown>
  ) => props;
};

export type Page = {
  readonly component: string;
  props: props;
  readonly url: string;
  readonly version: string;
};
export type Inertia = {
  readonly setViewData: (viewData: props) => Inertia;
  readonly shareProps: (sharedProps: props) => Inertia;
  readonly setStatusCode: (statusCode: number) => Inertia;
  readonly setHeaders: (headers: Record<string, string>) => Inertia;
  readonly render: (Page: Page) => Promise<Inertia>;
  readonly redirect: (url: string) => Context;
};
export const headers = {
  xInertia: 'x-inertia',
  xInertiaVersion: 'x-inertia-version',
  xInertiaLocation: 'x-inertia-location',
  xInertiaPartialData: 'x-inertia-partial-data',
  xInertiaPartialComponent: 'x-inertia-partial-component',
};
const inertia: (
  options: Options
) => Middleware<
  ParameterizedContext<DefaultState, DefaultContext>
> = function ({ version, html, flashMessages }) {
  return (ctx, next: Next) => {
    if (
      ctx.method === 'GET' &&
      ctx.headers[headers.xInertia] &&
      ctx.headers[headers.xInertiaVersion] !== version
    ) {
      ctx.status = 409;
      ctx.append(headers.xInertiaLocation, ctx.url);
      return;
    }

    let _viewData = {};
    let _sharedProps: props = {};
    let _statusCode = 200;
    let _headers = {};

    const Inertia: Inertia = {
      setViewData(viewData) {
        _viewData = viewData;
        return this;
      },

      shareProps(sharedProps) {
        _sharedProps = { ..._sharedProps, ...sharedProps };
        return this;
      },

      setStatusCode(statusCode) {
        _statusCode = statusCode;
        return this;
      },

      setHeaders(headers) {
        _headers = { ..._headers, ...headers };
        return this;
      },

      async render({ props, ...pageRest }) {
        const _page: Page = {
          ...pageRest,
          url: ctx.originalUrl || ctx.url,
          version,
          props,
        };
        log('rendering', _page);
        if (flashMessages) {
          this.shareProps(flashMessages(ctx));
        }

        const allProps = { ..._sharedProps, ...props };

        let dataKeys;

        if (
          ctx.headers[headers.xInertiaPartialData] &&
          ctx.headers[headers.xInertiaPartialComponent] === _page.component
        ) {
          dataKeys = ctx.headers[headers.xInertiaPartialData].split(',');
        } else {
          log(
            'partial requests without the name of the component return a full request',
            _page.component
          );
          log(
            'header component',
            ctx.headers[headers.xInertiaPartialComponent]
          );
          dataKeys = Object.keys(allProps);
        }

        // we need to clear the props object on each call
        const propsRecord: props = {};
        for await (const key of dataKeys) {
          log('parsing props keys', dataKeys);
          let value;
          if (typeof allProps[key] === 'function') {
            value = await (allProps[key] as () => unknown)();
            log('prop promise resolved', key);
          } else {
            value = allProps[key];
          }
          propsRecord[key] = value;
        }
        _page.props = propsRecord;
        log('Page props built', _page.props);

        if (ctx.headers[headers.xInertia]) {
          ctx.status = _statusCode;
          ctx.set({
            ..._headers,
            'Content-Type': 'application/json',
            [headers.xInertia]: 'true',
            Vary: 'Accept',
          });
          ctx.body = JSON.stringify(_page);
        } else {
          ctx.status = _statusCode;
          ctx.set({
            ..._headers,
            'Content-Type': 'text/html',
          });
          ctx.body = html(_page, _viewData);
        }
        return this;
      },

      redirect(url) {
        const statusCode = ['PUT', 'PATCH', 'DELETE'].includes(ctx.method)
          ? 303
          : 302;
        ctx.status = statusCode;
        ctx.set({ ..._headers, Location: url });
        return ctx;
      },
    };

    ctx.Inertia = Inertia;

    return next();
  };
};
export default inertia;


src/lib/expressAdapter.ts
import debug from 'debug';
import { RequestHandler, Response } from 'express';

const log = debug('inertia-node-adapter:express');

type props = Record<string | number | symbol, unknown>;

export type Options = {
  readonly enableReload?: boolean;
  readonly version: string;
  readonly html: (page: Page, viewData: props) => string;
  readonly flashMessages?: (req: unknown) => props;
};

export type Page = {
  readonly component: string;
  props: props;
  readonly url: string;
  readonly version: string;
};
export type Inertia = {
  readonly setViewData: (viewData: props) => Inertia;
  readonly shareProps: (sharedProps: props) => Inertia;
  readonly setStatusCode: (statusCode: number) => Inertia;
  readonly setHeaders: (headers: Record<string, string>) => Inertia;
  readonly render: (Page: Page) => Promise<Response>;
  readonly redirect: (url: string) => Response;
};
export const headers = {
  xInertia: 'x-inertia',
  xInertiaVersion: 'x-inertia-version',
  xInertiaLocation: 'x-inertia-location',
  xInertiaPartialData: 'x-inertia-partial-data',
  xInertiaPartialComponent: 'x-inertia-partial-component',
  xInertiaCurrentComponent: 'x-inertia-current-component',
};
const inertiaExpressAdapter: (options: Options) => RequestHandler = function ({
  version,
  html,
  flashMessages,
  enableReload = false,
}) {
  return (req, res, next) => {
    if (
      req.method === 'GET' &&
      req.headers[headers.xInertia] &&
      req.headers[headers.xInertiaVersion] !== version
    ) {
      return req.session.destroy(() => {
        res.writeHead(409, { [headers.xInertiaLocation]: req.url }).end();
      });
    }

    let _viewData = {};
    let _sharedProps: props = {};
    let _statusCode = 200;
    let _headers: Record<string, string | string[] | undefined> = {};

    const Inertia: Inertia = {
      setViewData(viewData) {
        _viewData = viewData;
        return this;
      },

      shareProps(sharedProps) {
        _sharedProps = { ..._sharedProps, ...sharedProps };
        return this;
      },

      setStatusCode(statusCode) {
        _statusCode = statusCode;
        return this;
      },

      setHeaders(headers) {
        _headers = { ...req.headers, ..._headers, ...headers };
        return this;
      },
      async render({ props, ...pageRest }) {
        const _page: Page = {
          ...pageRest,
          url: req.originalUrl || req.url,
          version,
          props,
        };
        log('rendering', _page);
        if (flashMessages) {
          log('Flashing messages');
          this.shareProps(flashMessages(req));
        }
        if (enableReload) {
          log('Setting session for reloading components');
          req.session.xInertiaCurrentComponent = pageRest.component;
        }
        const allProps = { ..._sharedProps, ...props };

        let dataKeys;
        const partialDataHeader = req.headers[headers.xInertiaPartialData];
        if (
          partialDataHeader &&
          req.headers[headers.xInertiaPartialComponent] === _page.component &&
          typeof partialDataHeader === 'string'
        ) {
          dataKeys = partialDataHeader.split(',');
        } else {
          log(
            'partial requests without the name of the component return a full request',
            _page.component
          );
          log(
            'header partial component',
            req.headers[headers.xInertiaPartialComponent]
          );
          dataKeys = Object.keys(allProps);
        }

        // we need to clear the props object on each call
        const propsRecord: props = {};
        for await (const key of dataKeys) {
          log('parsing props keys', dataKeys);
          let value;
          if (typeof allProps[key] === 'function') {
            value = await (allProps[key] as () => unknown)();
            log('prop promise resolved', key);
          } else {
            value = allProps[key];
          }
          propsRecord[key] = value;
        }
        _page.props = propsRecord;
        log('Page props built', _page.props);

        if (req.headers[headers.xInertia]) {
          res
            .status(_statusCode)
            .set({
              // ...req.headers,
              ..._headers,
              'Content-Type': 'application/json',
              [headers.xInertia]: 'true',
              Vary: 'Accept',
            })
            .send(JSON.stringify(_page));
          log(`sent response with headers`);
          log(res.getHeaders());
        } else {
          log('Sending the default html as no inertia header is present');
          res
            .status(_statusCode)
            .set({
              ...req.headers,
              ..._headers,
              'Content-Type': 'text/html',
            })
            .send(html(_page, _viewData));
        }
        return res;
      },

      redirect(url) {
        const statusCode = ['PUT', 'PATCH', 'DELETE'].includes(req.method)
          ? 303
          : 302;
        res.redirect(statusCode, url);
        log(`Redirecting to ${req.method} ${url}`);
        return res;
      },
    };

    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    req.Inertia = Inertia;

    return next();
  };
};
export default inertiaExpressAdapter;
